### PR TITLE
store: forbid collecting info if enable-collect-execution-info disabled

### LINKED ISSUE #31038: Runtime stats collection shouldn't collect stats if enable-collect-execution-info isn't enabled.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Start a cluster and running a workload with setting `enable-collect-execution-info` as false. Scrape the heap profile on tidb-server and we can see the following case:

image

If we try to use `list` to see the detail in function, we can find that many little memory objects make consuming lots of memory due to collecting stats info event if we disabled it.

image

image

### 2. What did you expect to see? (Required)

There shouldn't exist so much memory consuming.

### 3. What did you see instead (Required)

As above said

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: close #31038 

### What is changed and how it works?

forbid collecting info if enable-collect-execution-info disabled

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: store/copr/coprocessor.go [hunk 1]
```diff
@@ -980,11 +985,54 @@ func (worker *copIteratorWorker) handleCopResponse(bo *Backoffer, rpcCtx *tikv.R
...
+}
+
+func (worker *copIteratorWorker) handleCollectExecutionInfo(bo *Backoffer, rpcCtx *tikv.RPCContext, resp *copResponse) {
+	defer func() {
+		worker.kvclient.Stats = nil
+	}()
+	if !worker.enableCollectExecutionInfo {
+		return
+	}
+	failpoint.Inject("fix-31038", func(val failpoint.Value) {
```
Comment (ichn-hu) [inline store/copr/coprocessor.go:1027]: please give the failpoint and the test case a more intuitive name, and optionally give the link to the issue.

File: kv/kv.go [hunk 2]
```diff
@@ -238,12 +238,20 @@ type Transaction interface {
 // Client is used to send request to KV layer.
 type Client interface {
```
Comment (ichn-hu) [inline kv/kv.go:239]: does br or lighting use this interface? I am not sure if there are other code replying on this interface, do you carefully checked them all?

**Review by ichn-hu — CHANGES_REQUESTED** [review]
btw, will TiKV and TiFlash waste resource collecting stats if the stats collection is disabled?

File: kv/kv.go (hunk 2 above)
Comment (Yisaer) [reply kv/kv.go:239]: As `BR` and `lightning` are merged in tidb repo now, I think it's ok.

**Comment by Yisaer** [comment]
> btw, will TiKV and TiFlash waste resource collecting stats if the stats collection is disabled?

As there is no memory consuming problem exposed in the issue case, I think it doesn't matter now.

**Review by ichn-hu — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

File: executor/executor_test.go [hunk 3]
```diff
@@ -9751,3 +9751,17 @@ func (s *testSerialSuite) TestIssue30971(c *C) {
 		c.Assert(fields, HasLen, test.fields)
 	}
 }
+
+func (s *testSerialSuite) TestFix31038(c *C) {
```
Comment (JmPotato) [inline executor/executor_test.go:9755]:
```suggestion
// Details at https://github.com/pingcap/tidb/issues/31038
func (s *testSerialSuite) TestFix31038(c *C) {
```

**Review by JmPotato — APPROVED** [review]

**Pull request merged**